---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE, warning=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)

library(admixturegraph)
suppressPackageStartupMessages(library(dplyr, quietly = TRUE))
suppressPackageStartupMessages(library(ggplot2, quietly = TRUE))
suppressPackageStartupMessages(library(ggthemes, quietly = TRUE))
suppressPackageStartupMessages(library(knitr, quietly = TRUE))
suppressPackageStartupMessages(library(neldermead, quietly = TRUE))
```

# admixturegraph: Admixture Graph Manipulation and Fitting

The package provides functionality to analyse and test admixture graphs against the f statisticsdescribed in the paper [Ancient Admixture in Human History](http://tinyurl.com/o5a4kr4), Patternson *et al.*, Genetics, Vol. 192, 1065--1093, 2012.

The f statistics --- f2, f3, and f4 --- extract information about correlations between gene frequencies in different populations (or single diploid genome samples), which can be informative about patterns of gene flow between these populations in form of admixture events. If a graph is constructed as a hypothesis for the relationship between the populations, equations for the expected values of the f statistics can be extracted, as functions of edge lenghs — representing genetic drift — and admixture proportions.

This package provides functions for extracting these equations and for fitting them against computed f statistics. It does not currently provide functions for computing the f statistics — for that we refer to the [ADMIXTOOLS](https://github.com/DReichLab/AdmixTools) software package.

## Example

Below is a quick example of how the package can be used. The example uses data
from polar bears and brown bears with a black bear as outgroup. The BLK sample
is the black bear, the PB and AK populations are polar bears, and the rest
are brown bears. There are two populations of so-called ABC bears, ABC_A and
ABC_BC, a population of Yellow Stone bears, YB, and a population of European
brown bears, EBB.

Using ADMIXTOOLS I have computed f4 statistics --- there called D statistics
--- and have the results:

```{r}
data(bears)
bears
```

The D column is the f4(W,X;Y,Z) statistic and the Z column is the Z-values
obtained from a blocked jacknife (see Patterson *et al.* for details).

From the statistics we can see that the ABC bears are closer related to the
polar bears compared to the other brown bears and that the Yellow Stone bears
are closer to the polar bears than the European bears, and also that the two
populations of ABC bears are not equally pulled towards the polar bears.

We can explain this by several waves of admixture from ancestral polar bears
into brown bears, so a first attempt of a graph can be constructed as this:

```{r first_graph}
leaves <- c("BLK", "PB", "AK", "ABC_BC", "ABC_A", "YB", "EBB") 
inner_nodes <- c("R", "a", "b", "c", "d", "e", "f", "g", "h",
                 "abc_bc", "G", "E")

edges <- parent_edges(c(edge("BLK", "R"),
                        edge("PB", "h"),
                        edge("AK", "h"),
                        edge("ABC_BC", "abc_bc"),
                        admixture_edge("abc_bc", "f", "g"),
                        edge("ABC_A", "g"), 
                        edge("YB", "e"),
                        edge("EBB", "c"),
                        edge("h", "f"),
                        edge("f", "d"),
                        edge("g", "G"),
                        admixture_edge("G", "d", "e"),
                        edge("d", "b"),
                        edge("e", "E"),
                        admixture_edge("E", "b", "c"),
                        edge("b", "a"),
                        edge("c", "a"),
                        edge("a", "R")))
 

admixtures <- admixture_proportions(c(admix_props("abc_bc", "f", "g", "alpha"),
                                      admix_props("G", "d", "e", "beta"),
                                      admix_props("E", "b", "c", "gamma")))
                                
bears_graph <- agraph(leaves, inner_nodes, edges, admixtures)
plot(bears_graph, show_inner_node_labels = TRUE, show_admixture_labels = TRUE)
```

The graph makes predictions on how the f4 statistics should look, in particular
it allows us to predict the signs of the f4 statistics.

```{r}
add_graph_f4_sign(bears, bears_graph)
```

Here we see that it generally predicts the signs well, but I made a mistake in
the structure of the ABC bears --- the A versus BC statistics compared to polar
bears have the wrong sign --- so we can try to update the graph.

```{r second_graph}
leaves <- c("BLK", "PB", "AK", "ABC_A", "ABC_BC", "YB", "EBB") 
inner_nodes <- c("R", "a", "b", "c", "d", "e", "f", "g", "h",
                 "abc_a", "G", "E")

edges <- parent_edges(c(edge("BLK", "R"),
                        edge("PB", "h"),
                        edge("AK", "h"),
                        edge("ABC_A", "abc_a"),
                        admixture_edge("abc_a", "f", "g"),
                        edge("ABC_BC", "g"), 
                        edge("YB", "e"),
                        edge("EBB", "c"),
                        edge("h", "f"),
                        edge("f", "d"),
                        edge("g", "G"),
                        admixture_edge("G", "d", "e"),
                        edge("d", "b"),
                        edge("e", "E"),
                        admixture_edge("E", "b", "c"),
                        edge("b", "a"),
                        edge("c", "a"),
                        edge("a", "R")))
 

admixtures <- admixture_proportions(c(admix_props("abc_a", "f", "g", "alpha"),
                                      admix_props("G", "d", "e", "beta"),
                                      admix_props("E", "b", "c", "gamma")))
                                
bears_graph <- agraph(leaves, inner_nodes, edges, admixtures)
plot(bears_graph, show_inner_node_labels = TRUE, show_admixture_labels = TRUE)

```

Which now predicts the signs better.

```{r}
add_graph_f4_sign(bears, bears_graph)
```

The way the signs are predicted is by extracting the equations for the f4
statistics that the graph implies: For each quartet of leaves we can extract
an equation for the corresponding f4 statistics --- an equation in the
edge lenghts and admixture proportions --- and if this equation only have positive
values we know that the sign must be positive, if it only has negative values we
know that it must be negative, and if it constant zero we know it must be zero.

For example

```{r}
sf4(bears_graph, "BLK", "PB", "ABC_A", "EBB")
```

only has negative terms so we know it must be negative.

In general we will not always have only positive or negative terms, in which case
we cannot this simply predict the sign for f4 statistics. If this is the case
we need to set the parameters of the graph --- the edge lengths and admixture 
proportions --- to get the sign, and in that case we can also predict the
numerical value of the f4 statistics from the graph.


## Fitting a graph to data

If you have the *neldermead* package installed you can also fit graph parameters to data.
This is done using the *fit_graph* function

```{r, warning=FALSE}
fit <- fit_graph(bears, bears_graph)
fit
```

The object it returns contains an environment that contains the fitted parameters and a data
frame containing the original data together with an extra column, graph_f4, containing the fitted
values.

You can get the fitted values by calling the *summary* function.

```{r}
summary(fit)
```

This function also returns the fitted values as a list, so you can assign the result to an object
if you need to access it later.

You can also get the fitted parameters using the generic *coef* or *coefficients* funcions

```{r}
coef(fit)
```

To get the fitted predictions, together with the data used for fitting, use the *fitted* function.

```{r}
fitted(fit)
```